Try: Add initial setting options render

joshuatf opened the pull request:
* Adds initial setting options rendering for `text|password|select|multiselect|number|textarea`
* Adds data store selector to get setting options

This PR requires follow-ups and DOES NOT:

* Address insufficient preloading of "General" settings
* Handle option styling
* Rendering HTML segments in options
* Option changes/state and saving

### Screenshots

<img width="552" alt="Screen Shot 2021-01-19 at 5 08 47 PM" src="https://user-images.githubusercontent.com/10561050/105099330-82fb9300-5a79-11eb-8d00-2d5a83fcf679.png">


### Detailed test instructions:

1. Navigate to any settings page that uses the Settings API  to create controls (except General due to pre-existing preloaded settings issues)
2. Check that the above Setting types render (unstyled)
3. Smoke test other areas in the app to make sure no setting regressions have occurred

joshuatf committed:
Retrieve settings from data store

joshuatf committed:
Store all setting options in data store

joshuatf committed:
Add NumberControl to experimental package

joshuatf committed:
Add initial settings render

louwie17 commented on client/settings/setting-option.js:
```diff
@@ -0,0 +1,106 @@
+		return null;
+	}
+
+	const renderControl = () => {
```
I don't know how many more field types we might get, and or customizations that can happen when rendering a setting component
If so, what I have done in the past is to create some kind of registry (look up table), where you register the field types with conditions. Either just `type` or more if needed.
It would return the component, and every component would take the same props. It could allow 3PDs to register their own customer setting definition, and use it to render something.

Kinda like this (doing it on the fly, this is very basic):
```
const SettingComp = SettingRegistry.get(type);

return <SettingComp 
  value={ value }
  onChange={ onChange }
  placeholder={ placeholder }
  type={ type }
   />;
```
Registry:
```
const text = {
  type: 'text',
  component: TextControl // or a custom one
  // anything else that could be of relevance.
};
SettingRegistry.register(text);
```

joshuatf replied:
This is a good idea, but this is all of the types with the exception of HTML sections.  We will allow custom components, but plan to do so using `SlotFill`.

louwie17 replied:
Nice, then my above suggestion would be overkill.
Just didn't want to end up at a point where we have a bunch of different setting types and use cases that could make a mess, and could cause for a lot of bugs.

louwie17 reviewed (commented):
Left a couple comments as food for thought :) I just looked at these changes, not the one's already one in `add/settings-client`.
I did notice one bug as mentioned below.

joshuatf commented on packages/data/src/settings/selectors.js:
```diff
@@ -18,6 +18,30 @@ export const getSettings = ( state, group ) => {
 		return settings;
 	}
+
+	settingIds.forEach( ( id ) => {
```
@louwie17 Can you send me an example of where you've done this?  I'm not quite following.

louwie17 replied:
Hmm not sure if I have an exact example, both of these PRs (one of mine and of Jeff's) address the new object reference (re-rendering thing): #5986 & #5759 

> We might want to separate the value from the setting definition, and pass the value further down in a useSelect.

I was thinking of something like this (this is psuedo code though):
```
function SettingsPage() {
  // Just the field definitions (what type, options, etc...) that will always stay the same.
  const fieldDefinitions = useSelect(() =>
    return getSettingFieldsDefinition();
 });
  return fieldDefinitions.map(def => <SettingOption ...def />);
}
```

Then in a field component:
```
function SettingOption({name, type }) {
 // the value of the input field
  const settingValue = useSelect(() => {
      return getSettingValue(name);
  });
}
```

joshuatf commented on client/settings/setting-option.js:
```diff
@@ -0,0 +1,106 @@
+			);
+		}
+
+		if ( type === 'password' ) {
+			return (
+				<TextControl
+					value={ value }
+					onChange={ onChange }
+					placeholder={ placeholder }
+				/>
+			);
+		}
+
```
It was, thanks for catching this.

joshuatf committed:
Handle PR feedback

joshuatf commented on packages/data/src/settings/selectors.js:
```diff
@@ -18,6 +18,30 @@ export const getSettings = ( state, group ) => {
+};
+
+const getSettingValues = ( settings ) => {
+	return settings.reduce( ( resource, setting ) => {
```
Ah, this is related to the general settings I mentioned above.  Was hoping to tackle in a follow-up, but since this portion is not behind a feature flag, I updated it to preload all general settings for this PR and may attempt to reduce some of this preloaded data in a follow-up.

>Address insufficient preloading of "General" settings

louwie17 replied:
Aaah sorry, I mis understood, I thought you were referring to the `General` settings page.
Thanks for fixing it anyway though :) given this is being merged into `add/settings-client` I assume it's not going into `main` just yet?

joshuatf commented:
Thanks for the review @louwie17!  Addressed your feedback and left one comment above in regards to https://github.com/woocommerce/woocommerce-admin/pull/6101#discussion_r561234772

louwie17 reviewed (approved):
@joshuatf left a reply to the comment you mentioned, otherwise things look good to me 👍  Thanks for fixing the home page anyway, although I misunderstood:

> Address insufficient preloading of "General" settings

samueljseay commented:
Does this one need a changelog entry?
